include/swift/AST/Attr.def

```diff
@@ -552,7 +552,18 @@ CONTEXTUAL_SIMPLE_DECL_ATTR(_resultDependsOn, ResultDependsOn,
 CONTEXTUAL_SIMPLE_DECL_ATTR(transferring, Transferring,
   OnParam | DeclModifier | UserInaccessible | NotSerialized | ABIBreakingToAdd | ABIBreakingToRemove | APIBreakingToAdd | APIStableToRemove,
   157)
-
+CONTEXTUAL_SIMPLE_DECL_ATTR(_copy, CopyLifetimeDependence,
+  OnFunc | DeclModifier | LongAttribute | AllowMultipleAttributes | ABIBreakingToAdd | ABIStableToRemove | APIBreakingToAdd | APIStableToRemove,
+  158)
+CONTEXTUAL_SIMPLE_DECL_ATTR(_consume, ConsumeLifetimeDependence,
+  OnFunc | DeclModifier | LongAttribute | AllowMultipleAttributes | ABIBreakingToAdd | ABIStableToRemove | APIBreakingToAdd | APIStableToRemove,
+  159)
+CONTEXTUAL_SIMPLE_DECL_ATTR(_borrow, BorrowLifetimeDependence,
+  OnFunc | DeclModifier | LongAttribute | AllowMultipleAttributes | ABIBreakingToAdd | ABIStableToRemove | APIBreakingToAdd | APIStableToRemove,
+  160)
+CONTEXTUAL_SIMPLE_DECL_ATTR(_mutate, MutateLifetimeDependence,
+  OnFunc | DeclModifier | LongAttribute | AllowMultipleAttributes | ABIBreakingToAdd | ABIStableToRemove | APIBreakingToAdd | APIStableToRemove,
+  161)
```

Comment:
I don't think the contextual attributes create Attr types. They are created only for DECL_ATTR and since they are handled with a switch, there would be several compiler errors if they existed.

In sema I raise an error if `_copy` etc are found in other positions.

Reply:
Well, I'm pretty sure it does, they are typedef of `SimpleDeclAttr<XXX>`, but still. 

`CONTEXTUAL_SIMPLE_DECL_ATTR` is defaulted to `SIMPLE_DECL_ATTR` and `SIMPLE_DECL_ATTR` is defaulted to `DECL_ATTR`.
https://github.com/apple/swift/blob/bc86b8db30c0e7777c2c25e97e0a5eec76d52926/include/swift/AST/Attr.def#L26-L34
`DeclAttrKind` are created here:
https://github.com/apple/swift/blob/bc86b8db30c0e7777c2c25e97e0a5eec76d52926/include/swift/AST/AttrKind.h#L133-L137
`typedef` are created here. 
https://github.com/apple/swift/blob/bc86b8db30c0e7777c2c25e97e0a5eec76d52926/include/swift/AST/Attr.h#L495-L498

> since they are handled with a switch, there would be several compiler errors if they existed.

Yes, that's why you added this in TypeCheckAttr.cpp.
```
  IGNORED_ATTR(CopyLifetimeDependence)
  IGNORED_ATTR(ConsumeLifetimeDependence)
  IGNORED_ATTR(BorrowLifetimeDependence)
  IGNORED_ATTR(MutateLifetimeDependence)
```
Other than that, we don't switch `DeckAttrKind` many places, they are handled by `default: ` or `SIMPLE_DECL_ATTR` cases are handled by meta programming like:
https://github.com/apple/swift/blob/5feb9988a5739f77b903d83b794cef1e93c5fb7b/lib/AST/Attr.cpp#L1012-L1014

My question is are they actually needed? If so, could you explain why?

Reply:
These are modifiers we want to add on a type. Just like consuming, borrowing parameter modifiers. The difference between this and other ownership modifiers are that these are on the result type.

Reply:
Expressing them as attributes provides an easy way to represent them in the ast, and all other related infrastructure like `applyAttributesOnType` etc

Reply:
My misunderstanding about the DECL_ATTR and CONTEXTUAL_SIMPLE_DECL_ATTR was because I wasn't  led to define my own type in `Attr.cpp` like `DynamicReplacementAttr` etc

Reply:
Not sure I'm fully following this discussion properly, but as far as I understand it @rintaro's point is that these aren't actually used by anything. They're either ignored specifically or ignored in the default branches where we switch over the attribute kinds.

`applyAttributesToType` doesn't use them, it uses:
```
 if (!lifetimeDependenceSpecifiers.empty()) {
    ty = LifetimeDependentReturnTypeRepr::create(p.Context, ty,
                                                 lifetimeDependenceSpecifiers);
  }
```

And `lifetimeDependenceSpecifiers` are `LifetimeDependenceSpecifier` which get created by `parseLifetimeDependenceSpecifiers`.

Are we missing a use anywhere? Otherwise this is actually actively harmful, as it will allow parsing them as a modifier which shouldn't be valid.

Reply:
@rintaro explained that I don't need to represent them as `CONTEXTUAL_SIMPLE_DECL_ATTR ` just for contextual keywords. So I removed it in my latest revision.